File: core/src/main/java/heast/core/network/packets/c2s/auth/SignupC2SPacket.java
package heast.core.network.packets.c2s.auth;

import heast.core.network.Packet;
import heast.core.network.PacketBuf;
import heast.core.network.listeners.ServerAuthListener;

public final class SignupC2SPacket implements Packet<ServerAuthListener> {

    private final String username;
    private final String email;
    private final String password;

    public SignupC2SPacket(String username, String email, String password) {
        this.username = username;
        this.email = email;
        this.password = password;
    }

    public SignupC2SPacket(PacketBuf buf) {
        this.username = buf.readString();
        this.email = buf.readString();
        this.password = buf.readString();
    }

    @Override
    public void write(PacketBuf buf) {
        buf.writeString(username);
        buf.writeString(email);
        buf.writeString(password);
    }

    @Override
    public void apply(ServerAuthListener listener) {
        listener.onSignup(this);
    }

    public String getUsername() {
        return username;
    }

    public String getEmail() {
        return email;
    }

    public String getPassword() {
        return password;
    }
}

File: Core/Network/Packets/C2S/SignupC2SPacket.cs
using Core.Network.Listeners;

namespace Core.Network.Packets.C2S;

public partial class SignupC2SPacket : AbstractPacket
{
    public SignupC2SPacket(string username, string email, string password)
    {
        Username = username;
        Email = email;
        Password = password;
    }

    public string Username { get; }
    public string Email { get; }
    public string Password { get; }

    public override void Apply(IPacketListener listener)
    {
        if (listener is IServerAuthListener authListener)
        {
            authListener.OnSignup(this);
        }
    }
}